### DISCUSSION TITLE
Anycubic Photon Mono M7 Pro format

### DISCUSSION THREAD

**Discussion by SirKliv** [sn4k3/UVtools#892] [Ideas]
Hello!
Is there any possibility to add new format support for [Anycubic Photon Mono M7 Pro](https://wiki.anycubic.com/en/resin-3d-printer/photon-mono-m7-pro)?
Sliced file suffix: pwsz
Zip image format: pwszImg
This format is supported in the Anycubic Photon Workshop slicer since v3.3.2
Thanks in advance!

**Comment by sn4k3** [comment]
No ETA to support that.
Also discussed here: https://www.facebook.com/groups/uvtools/posts/1115738076202361/

**Comment by kozross** [comment]
@sn4k3 - what information would you need in order to support this format?

**Reply by sn4k3** [reply]
The layer image compression / encryption

**Comment by nalimcos** [comment]
Hello @sn4k3 ! I just bought one of these, and happen to be a developer (no experience with file formats though).

I have taken a look at a file sliced by Chitubox 2.2 (which I have already used to successfully print with this machine) and here is what I found:

Overall container: .pwsz -> as expected, it's just a zip file
![image](https://github.com/user-attachments/assets/e42b0c80-889e-4bb3-9151-27f0c9e967d8)

Layer images:
layer_xx.pw0Img
Screen resolution is 13312*5120, so I'd expect 68 157 440 bytes for a raw image at 1 byte per pixel
The files sizes vary around half of that though, so there is some compression there. Some form of RLE? 

File beginning:
![image](https://github.com/user-attachments/assets/05106f64-079f-4e16-b1d8-699afccf7f19)

Transition from '0f ff' repetition (looks suspiciously like RLE4EncodingLimit in your code, but I haven't checked whether that's relevant) to what seems to represent the tiny test object in the center of the build plate:
![image](https://github.com/user-attachments/assets/664ad74f-a587-4fbb-9c35-da532667fecf)

File ending:
![image](https://github.com/user-attachments/assets/5a3de91a-7145-442f-bb0c-675487cfc092)

Here is the entire pwsz file (.zip added so github would let me attach it):
[test.pwsz.zip](https://github.com/user-attachments/files/17936049/test.pwsz.zip)

How much development time do you think it would take you to get the file format added (basic support at least), along with PrusaSlicer integration? I'd be glad to either contribute financially to move this up the priority list (my budget would probably not cover outright hiring you to get it done :laughing: ), or to do it myself if you can guide me to some extent as I'm unfamiliar with the project.

**Comment by rncwind** [comment]
Not an M7 pro but a standard M7 owner here.
File format seems to mostly be the same after slicing in lychee, except with a ``pm7`` extension instead of ``pwsz``.

Basically the same zip file contents listing after renaming ``pm7`` to ``zip``
![image](https://github.com/user-attachments/assets/373090da-70c9-492f-9b3c-9dc339344077)
Layer images are also ``.pw0Img`` files.
![image](https://github.com/user-attachments/assets/8d77d6f8-3d9c-4b3c-b05e-974eb05342df)

Makes sense since they're pretty much the same machine internals wise.
Hopefully this can be of some little help!

**Comment by nalimcos** [comment]
Quite a bit of the necessary work has actually already been done, uvtools can read some Photon M7 files when the debug flag is set, and it didn't take much fiddling around to read M7 Pro files. (It looks like non-square pixels aren't really handled by the preview at the moment though!)

![image](https://github.com/user-attachments/assets/3d86957c-02eb-41c0-8cbc-1e95eee3945a)

Apparently, layer images can be either .pwszImg or .pw0Img. Also, my chitubox input file had key_image_format pwszImg in anycubic_photon_resins.pwsp while the layers were .pw0Img, so the extension should be checked against the contents of the folder, not the metadata.
![image](https://github.com/user-attachments/assets/5b03c1b4-1808-4663-b249-4616da934f75)

Export to M7 Pro was not a great success: layer_images was entirely missing, haven't searched for the reason yet.

There seems to be some work needed on the settings (anycubic_photon_resins.pwsp), not sure whether it is serviceable as is, but the heated vat is not supported yet any rate; not sure about the rest.

**Reply by sn4k3** [reply]
> Apparently, layer images can be either .pwszImg or .pw0Img. Also, my chitubox input file had key_image_format pwszImg in anycubic_photon_resins.pwsp while the layers were .pw0Img, so the extension should be checked against the contents of the folder, not the metadata.

Using a slicer other than the original for developing is often an error. In past versions Chitubox had bugs in anycubic file format implementation which broke UVtools validations (They fixed). This is why you always want to use original slicer to generate the files and compare with.
About metadata, my experience tells me that's only informative, printer don't care about it most of the time as firmware is already defined to follow only a way, so you can declare whatever you want as long the data it reads are in the expected format.

Please slice only with Anycubic slicer and confirm the data.

**Reply by nalimcos** [reply]
The thought had crossed my mind, but I got lazy as it would have required rebooting into Windows or fiddling with WINE, while I already had Chitubox installed. I will compare the formats when I get a chance. We did, at any rate, get some information on one variant of what the printer accepts (and what one slicer currently outputs).

I suspect there is an actual temperature control somewhere, as the printer ignores what I set in its interface and just preheats to 25C (which I see several times in the file) - I will do some testing.

**Reply by sn4k3** [reply]
Here two files sliced by me in anycubic slicer, both produce pwsz images

[fish_fossilz.pwsz.zip](https://github.com/user-attachments/files/17967549/fish_fossilz.pwsz.zip)
[fish_fossilz.pm7.zip](https://github.com/user-attachments/files/17967548/fish_fossilz.pm7.zip)

**Reply by kozross** [reply]
@nalimcos - what WINE setup do you use for running Photon Workshop? I've tried it a few times, and it never worked for me.

**Reply by sn4k3** [reply]
> @nalimcos - what WINE setup do you use for running Photon Workshop? I've tried it a few times, and it never worked for me.

For testing purposes you can mount a window VM, much more reliable than wine

**Reply by tslater2006** [reply]
@sn4k3 I just wanted to provide these files. These are made from Photon Workshop 3.4.2 and are files for the M7 Max and M7 Pro. Notably, while they are the AnycubicZipFile format, and `anycubic_photon_resins.pwsp` states `"key_image_format": "pwszImg",` if you look in the `layer_images` directory this files are annoyingly suffixed with `.pw0Img`

![image](https://github.com/user-attachments/assets/7d476805-3ab0-48f6-87a3-423048477662)

[Cone_Files.zip](https://github.com/user-attachments/files/18001437/Cone_Files.zip)

**Reply by tslater2006** [reply]
Also, those *.pw0Img files correctly decode with the DecodeLayerRLE method. For testing I just replaced 
`entry = inputFile.GetEntry($"layer_images/layer_{layerIndex}.pwszImg");` with `entry = inputFile.GetEntry($"layer_images/layer_{layerIndex}.pw0Img");` locally to see if it would load them and it does. (i had to change the pm7m extension of course for that one)

Also not quite on topic but adjacent: Is my reading of the current state correct in that going from some other file format (like VDT) to AnycubicZipFile is not currently supported? It seems to be missing a way to encode the images to RLE and save them to the ZIP. Is this something you would be interested in a PR for?

**Reply by sn4k3** [reply]
> These are made from Photon Workshop 3.4.2 and are files for the M7 Max and M7 Pro.

With same version I only get pw0 images with M7 Max, the two others generate pwsz images.

> Also, those *.pw0Img files correctly decode with the DecodeLayerRLE method

Yes, those pw0 will decode and encode correctly with pw0 methods as they are the same enconding.

> Is my reading of the current state correct in that going from some other file format (like VDT) to AnycubicZipFile is not currently supported? It seems to be missing a way to encode the images to RLE and save them to the ZIP.

The format is not completed, it was created in debug mode only for development, it's not available into release.
For now, the pwsz image is what is missing...

**Comment by ejongejans** [comment]
For what it's worth, I've been looking at the pwszImg format binary now for some minutes. It seems to me that: 

- it could be some kind of vector format. When exporting a simple cube, there is way to little data to be an RLE scheme. Because on this high screen resolution you would have many similar rows with black-white-black. At least a 100 or so. Not seeing that. 

- i first thought maybe they index every row again, so similar row's (black-white (cube)-black) get repeated. But then I tried a diamond shape. (many different rows) And the size of the binary files is only a little bigger..  

- there are some patterns with {== and [-- which seem somewhat human readable

- i scanned it a couple of times for float's, there are a lot of them. But i can't make sense of the values yet.. so it could be coincidence.. It doesn't seem direct coordinates of my mesh. Perhaps still encrypted somehow..

- saving 2x the same file yields the same. So no random keys?

10mm cube centered: 
![image](https://github.com/user-attachments/assets/3a5075ba-0701-4e1b-9e2c-e7ba083b85e8)

3x 10mm cube dragged somewhere random: 
![image](https://github.com/user-attachments/assets/690b6289-ce44-4d2e-944d-0e8d814681cf)

**Reply by ejongejans** [reply]
I have a 10mm cube now. When putting values to big endian I can read floats 4.75, -5.25, 5 and 5 at various locations. This is probably the way to read it.. 

![image](https://github.com/user-attachments/assets/efbc987d-8ccb-49ac-b3e2-584b825755eb)

**Reply by ejongejans** [reply]
continued on this.. here is a cube of 10 x 10 mm.. probably a bit off center. 
It seems it's writing out line segments in x1,y1->x2,y2 in big endian floating point format. There is 1 byte in between each line segment.
However it seems the curve is not a closed curve.. which would mean the printer not only rasterizes but also sorts the segments.. 


![image](https://github.com/user-attachments/assets/afdc4f03-b6ef-43b0-9f65-9d966c6f101e)

**Reply by ejongejans** [reply]
the 01 (3x) and 00 (1x) might be the directions of the segments.. no idea why you would do it like that..

**Comment by tslater2006** [comment]
This image format is RLE and actually already supported by UVTools in other
Anycubic formats. The reason it is broken for this printer is they
(Anycubic) named the files with a different extension. One that doesn't
match the actual file format -_-

**Reply by ejongejans** [reply]
So each of the 7 series machines (normal, max and pro) supports the PW0 RLE encoding?

**Reply by tslater2006** [reply]
What I'm saying is that *.pm7m files have images like this:
![image](https://github.com/user-attachments/assets/ce7c1e73-bb0b-4e05-97d0-633ca4f139f4)

But that file extension is *wrong* according to the settings in the pm7m file which indicates the file format is: 
![image](https://github.com/user-attachments/assets/f68aba21-d92a-479a-8d37-fe27635ebd84)

In fact the files *are* pwszImg format which is already supported (read only) by UVTools here:
https://github.com/sn4k3/UVtools/blob/7def1c1e8c0714e49e64f560d126280eb96f0068/UVtools.Core/FileFormats/AnycubicZipFile.cs#L536

If you change this line to be `entry = inputFile.GetEntry($"layer_images/layer_{layerIndex}.pw0Img");` then pm7m files load just fine.
https://github.com/sn4k3/UVtools/blob/7def1c1e8c0714e49e64f560d126280eb96f0068/UVtools.Core/FileFormats/AnycubicZipFile.cs#L830

In short, Anycubic says the format is one thing but then gives them the wrong extension. UVTools looks for the extension pwszImg, but for pm7m they decided to use pw0Img even though it is a pwszImg format... fun huh?

**Reply by tslater2006** [reply]
And somehow UVTools is supposed to just know that the file format is lying to it lol

**Reply by ejongejans** [reply]
yeah.. it's like a box of chocolates.. you never know what you are gonna get ;) 

But i'm looking specifically at the .pwszImg files; as i think these are used in the .pm7/.pwsz file. 
From what i read above they have not been documented yet. And that's why UVTools is not able to read/export to these other 2 machines. So probably the Max machine works; but that has half the resolution.. 

My guess is they went to vector format (storing vectors instead of RLE schemes) so save file size and rasterize on the printer.. as the resolutions and thus files just get to big..

**Reply by sn4k3** [reply]
I have not take a look into the pwsz more seriously but I will test that, I made 1px file to see the data.
But vector idea is not good if you use AA or blur. But the true is with the increasing resolution you need it less.

**Reply by ejongejans** [reply]
My guess is the machine rasterizes it then. So that could still mean AA being applied... depending how well it reads the settings :D

**Reply by ejongejans** [reply]
-xmin,ymin,xmax,ymax on top. 
-before that 200? (unsure) 
-dark purple, the 4 is the segment count. 
- other zero's i don't know.. 

![image](https://github.com/user-attachments/assets/516c3d11-ac7f-4f1e-9677-840981e555cc)

**Comment by Psychokiller1888** [comment]
I worked a lot on the pwszimg files. You are correct, but it's not xmin y min x max y max. The new format defines zones in the screen to minimize the file size. You have some Metadata in the first segment, that defines a global bounding box, with xmin ymin xmax ymax with some other data. That is defined after {===. It's like a fdm skirt to explain it. Then, you have blocks of data, after [===, that defines xstart, ystart, xend and yend. They are drawing a line as the base. Y minus goes up and X is mirrored. Data compression is tricky then

**Comment by sn4k3** [comment]
With all the information here I was able to decode all the pwsz format, you can find the definition [here](https://github.com/sn4k3/UVtools/blob/master/Scripts/010%20Editor/pwsz.bt).

![image](https://github.com/user-attachments/assets/791a11b2-a045-4ed5-be87-c1544df3a320)

Interesting notes:
- If you use AntiAliasing they turn in use pw0 instead, as of course line/contour for AA is very heavy and not worth, this means, if you choose no AA it will render pwsz otherwise pw0
- Can not find the meaning of Coordinate Unknown field, is either Zero or One, not sure when to use 1 or 0...

**Reply by ejongejans** [reply]
Thanks for your feedback. 

 the Coordinate Unknown field you have in the segments indicates the orientation of the segment (CW or CCW). The impression i got is that the slicer just slices all triangles and outputs the 2d segments intersecting the slice plane. It doesn't build the topology of the contours as this is more time consuming... 

The 1 or 0 probably follows from a 2d cross product with the xy of the face normal. Where material of the part should always be on the left. (CCW). Instead of flipping the coordinates they output a 0 or 1..  no idea why. Perhaps a hint for their rasterizer.. 
(it could also be looked at as an index which coordinate is first for that particular segment...  

Strictly speaking you don't need the direction of the segment to build the contour. But your search can be a bit quicker if you know all segments are in the correct orientation.

**Reply by ejongejans** [reply]
What tool is that you are using btw? I have an old hex analyser but somehow for float's it shows the wrong endianess.. as I wrote above 'big endian', everything was clearly little endian ;)

**Reply by ejongejans** [reply]
Perhaps the rasterizer is a fairly simple line fill.. 
- first is draw's all the segments with bresenhamms algoritm on the screens framebuffer
- then it fills from left to right
Would be interesting to test with an open contour... if you then get a line filling the screen untill the end...

**Reply by sn4k3** [reply]
Getting some progress:

![image](https://github.com/user-attachments/assets/351a6130-b324-4a89-acd2-4fdbc306a4fd)

> What tool is that you are using btw? I have an old hex analyser but somehow for float's it shows the wrong endianess.. as I wrote above 'big endian', everything was clearly little endian ;)

010 Editor

As line scan seems tricky I plan to ditch pwsz and always encode as pw0 😅

**Reply by ejongejans** [reply]
Looks very good! That seems enough of a proof that this is it. 

pw0 -> yep smart move i guess. 
Would all machines support the pw0 as well? I think the pw0 was pretty 'large' in file size right? So maybe the 2 machines with super high res screen only support pwsz?

**Reply by sn4k3** [reply]
Is a matter of test, but as I found, if you use AA the anycubic slicer export to pw0.... So I guess it's compatible.
The annoyance of pwsz is that they have jumps between objects lines

**Reply by sn4k3** [reply]
I may try doing a encode for PWSZ, if so anyone can test a sample file?

**Reply by Psychokiller1888** [reply]
Sure

**Reply by sn4k3** [reply]
> Sure

Whats your machine model?

But I will stick with pw0 for release, since I tested and my vector generation is larger than pw0 compression. (I don't get the perfect lines as I would from the orginal stl)

**Reply by Psychokiller1888** [reply]
As I said several times, I worked a lot of the slice export, as I'm using a M7 pro, that uses the pwszImg files. I'm slicing with Lychee and offered them help about the export aboht a month and a half ago and then found this GH issue a couple of days ago.

**Reply by sn4k3** [reply]
Try these two files: [fish_fossilz_30m_9ml_UVTOOLS.zip](https://github.com/user-attachments/files/18141879/fish_fossilz_30m_9ml_UVTOOLS.zip)
Print without VAT and a white paper sheet and observe images. One is encoded with PW0 other with PWSZ from UVTools encoder.

As additional note, I find PW0 better than new pwsz, not to say that pwsz doesn't support AA. 
PWSZ can only win on regular shapes on all file, even then not worth IMO.

Comparison (First is from AC slicer):

![image](https://github.com/user-attachments/assets/a8a5cdb0-bae9-4cab-8db9-df343807b047)

Shape:

![image](https://github.com/user-attachments/assets/af8d3c28-ddf7-42b5-9437-f8536287eb9f)

**Reply by Psychokiller1888** [reply]
Thank you, I will try this later, now time to go work. But....

I worked on the .pwsz slice files, for Lychee and helping them integrate the full ecosystem. The last part missing were those pwszImg files. I made hundreds of tests and exports, generation of images. I don't know, maybe you did not read some of my comments, so here some condensate

- Anycubic Photon Slicer doesn't allow you to turn of AA entirely, you always have some AA turned on. You can see it in their "Slicer", and they both export in pwszImg.
- Some parts of the pixel blocks in the pwszImg contain grey level information
- A 10x10 square slice is 68 octets
- pwszImg win over pws0 file because the former only adresses a part of the screen, where the later adresses the entire screen. It results, for a simple 10x10x10 cube stl of a file size 286 times larger using the later.
- Screenshot of a same 50x50x50 cube exported with Photon Slicer, one with AA, the other one """""without""""" AA, meaning, setting left to 1 in the slicer
![explorer_6VSGoZwsWX](https://github.com/user-attachments/assets/b8775622-e0d9-4f4d-964f-376357538492)

I have more, but I have to go to work now, so I'll report later

**Reply by sn4k3** [reply]
> Anycubic Photon Slicer doesn't allow you to turn of AA entirely, you always have some AA turned on. You can see it in their "Slicer", and they both export in pwszImg.

While they seem to use AA=4 (if you select AA 1 from 4) (In the json is possible to observe that they use a raster_aa of 4 but also in preview), the encoded line does not contain any grey pixel information, it may be just for a better raster and get fewer lines because AA=1 would cause more segmentation. If machine fw then does apply any AA to the raster lines (model) is another topic and not relevant to the format implementation.

AA1 vs AA4 equality proves that they use fixed AA within this levels:
![image](https://github.com/user-attachments/assets/84767b58-e76f-41b5-b4e2-05473e6b9c93)


From AA8 to AA16 anycubic start to output PW0 and warns about file size:

Even with AA8 and PW0 generated by anycubic, the file size is lower than pwsz 🤣
![image](https://github.com/user-attachments/assets/84236faf-1dba-4e22-8417-1504587cc81e)


> Some parts of the pixel blocks in the pwszImg contain grey level information

So far, I can only see colorless XY lines. If you have a sample with grey level information, please send me the sample.

> pwszImg win over pws0 file because the former only adresses a part of the screen, where the later adresses the entire screen. It results, for a simple 10x10x10 cube stl of a file size 286 times larger using the later.

I partial disagree. You win with regular shapes or huge shapes where a group of fill poly gives you advantage over rle. In most and real prints people will not print regular shapes that often and most are for the minis (many small areas), not to mention that supports and round shapes starts to kill compression as there are no arc support. 

On my previous screenshot you can notice that anycubic pwsz image is heavier than pw0 counterpart, using a corona model with supports, yields in half file size for pw0.
First layers you can observe that pwsz is smaller than pw0, and that's true due the shape of the raft (regular), if you start to scroll layers you notice the complexity of the model defeats pwsz, pw0 weight half for layer 700 as example:

![image](https://github.com/user-attachments/assets/733f9172-2554-4508-a746-0e9ac4e43d83)

Then I grab a mini model and scale to build plate, as expected pwsz give a slight of advantage because of the single huge area, still is only 3MB better than PW0, and my thought remains, there's no need for pwsz because the size benefits will likely not to happen or when does the difference is small. Machine may raster the pwsz faster, but this is also another topic which need testing.

![image](https://github.com/user-attachments/assets/fe2f2b88-9e23-43dc-a9ab-c30b176bcffa)

Then I duplicate the above test and slice two in the plate, as expected pwsz starts to lose!

![image](https://github.com/user-attachments/assets/eb52ceda-70bf-4172-b35c-bd59cdb21efd)

More complex objects will degrade the pwsz compression over pw0.
Is clear that pwsz will perform worse than pw0 in size for the common people. For who print functional parts it will benefit but who will care for a few MB's if printer perform the same?

## Some observations:

- UVtools have some line formats, some even support color information in line, but they are all bad in compression, they are worse than pwsz because they are either vertical or horizontal line (no diagonals)
- UVtools have vector-based file format (SVGX), which are true svg, yields better results than pwsz.

IMO pwsz is a failure, because at this stage they could have come with something better. eg: vectors (svg) which instead of lines can compress into single points and make contours, as so, instead a 17 byte line system they could have 6/8 byte contour system. That way a square would take `3 * 8 bytes = 24 bytes`, but currently it does in `4 * 17 bytes = 68 bytes`. The vector system would take 2.25x less file size, still without AA support.

## Things to consider:

If the machine fw would apply an AA to the pwsz raster then is a valid point to use PSWZ instead PW0, otherwise if they expose raw without AA post-process then is "useless" compared with PW0. This is because PW0 with no AA will compress as good as or better in many situations but if you add AA to PW0 it will increase size for a good margin, yet is possible to oberve some cases with AA=8 yields lower file size than PWSZ by anycubic slicer, so... 
This can be tested but will be very hard to notice 1 pixel wide AA. Maybe a phone with good zoom can make this conclusion?

**Comment by Psychokiller1888** [comment]
I made exports fron AC photon workshop in pwz with AA. I don't think you can turn off AA on their default slicer
